File: addTwoNumbers/addTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  var n1 = 0;
  var n2 = 0;
  l1.forEach((value, place) => {
    n1 += value * Math.pow(10, place);
  });
  l2.forEach((value, place) => {
    n2 += value * Math.pow(10, place);
  });
  var n3 = n1 + n2;
  var l3 = [];
  var place = 0;
  if (n3 === 0) {
    return [0];
  }
  while (Math.floor(n3 / Math.pow(10, place)) > 0) {
    l3.push(Math.floor(n3 / Math.pow(10, place) % 10));
    place++;
  }
  return l3;
};

var r1 = addTwoNumbers([2, 4, 3], [5, 6, 4]);
console.log(r1);

var r2 = addTwoNumbers([0], [0]);
console.log(r2);

var r3 = addTwoNumbers([9, 9, 9, 9, 9, 9, 9], [9, 9, 9, 9]);
console.log(r3);

File: addTwoNumbers/addTwoNumbers.java
package com.example.java;

public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode n1 = l1;
        ListNode n2 = l2;
        ListNode l3 = new ListNode();
        ListNode n3 = l3;
        boolean n1done = false;
        boolean n2done = false;
        int remainder = 0;

        while (!n1done || !n2done || remainder > 0) {
            int placeVal = 0;
            placeVal += remainder;
            remainder = 0;

            if (!n1done) {
                placeVal += n1.val;
                if (n1.next != null) {
                    n1 = n1.next;
                } else {
                    n1done = true;
                }
            }

            if (!n2done) {
                placeVal += n2.val;
                if (n2.next != null) {
                    n2 = n2.next;
                } else {
                    n2done = true;
                }
            }

            if (placeVal >= 10) {
                remainder = 1;
            }
            if (placeVal < 10) {
                n3.val = placeVal;
            } else {
                n3.val = placeVal - 10;
            }

            if (!n1done || !n2done || remainder > 0) {
                n3.next = new ListNode();
                n3 = n3.next;
            }
        }

        return l3;
    }
}
